src/resdk/resources/geneset.py

```diff
@@ -0,0 +1,161 @@
+"""Geneset resource."""
+import json
+import logging
+from urllib.parse import urljoin
+
+from .data import Data
+from .utils import get_collection_id
+
+
+class Geneset(Data):
+    """Resolwe Geneset resource.
+
+    :param resolwe: Resolwe instance
+    :type resolwe: Resolwe object
+    :param model_data: Resource model data
+
+    """
+
+    def __init__(self, resolwe, genes=None, source=None, species=None, **model_data):
+        """Initialize attributes."""
+        self.logger = logging.getLogger(__name__)
+
+        self._genes = None
+        if genes is not None:
+            self._genes = set(genes)
+
+        self._source = source
+        self._species = species
+
+        super().__init__(resolwe, **model_data)
+
+    @property
+    def genes(self):
+        """Get genes."""
+        if self._genes is None or len(self._genes) == 0:
+            self._genes = set()
+            if self.id and "geneset_json" in self.output:
+                url = urljoin(
+                    self.resolwe.url,
+                    "api/storage/{}".format(self.output["geneset_json"]),
+                )
+                response = self.resolwe.session.get(url, auth=self.resolwe.auth)
+                response = json.loads(response.content.decode("utf-8"))
+                self._genes = set(response["json"]["genes"])
+        return self._genes
+
+    @genes.setter
+    def genes(self, genes_list):
+        """Set genes."""
+        self._assert_allow_change("genes")
+        if genes_list is not None:
+            self._genes = set(genes_list)
+
+    @property
+    def source(self):
+        """Get source."""
+        if self._source is None and self.id and "source" in self.output:
+            self._source = self.output["source"]
+        return self._source
+
+    @source.setter
+    def source(self, new_source):
+        """Set source."""
+        self._assert_allow_change("source")
+        self._source = new_source
+
+    @property
+    def species(self):
+        """Get species."""
+        if self._species is None and self.id and "species" in self.output:
+            self._species = self.output["species"]
+        return self._species
+
+    @species.setter
+    def species(self, new_species):
+        """Set species."""
+        self._assert_allow_change("species")
+        self._species = new_species
+
+    def _assert_allow_change(self, field_name):
+        """Assert that this Geneset obj is not saved yet."""
+        if self.id:
+            msg = "Not allowed to change field {} after geneset is saved".format(
+                field_name
+            )
+            raise ValueError(msg)
+
+    def save(self):
+        """Save Geneset to the server.
+
+        If Geneset is already on the server update with save() from base class. Otherwise, create
+        a new Geneset by running process with slug "create-geneset".
+        """
+
+        if self.id:
+            super().save()
+        else:

```

Comment:
Since the new constructor can accept values of source species and genes as `None`, add a check that verifies that these values are not `None` at the time of Geneset creation.

Reply:
Otherwise process will fail.